xrtl/port/windows/ui/win32_control.cc

```diff
@@ -733,5 +743,190 @@ LRESULT Win32Control::WndProc(HWND hwnd, UINT message, WPARAM w_param,
+      wheel_delta = GET_WHEEL_DELTA_WPARAM(w_param);
+      break;
+    case WM_MOUSEMOVE:
+      // No action button for mouse move.
+      break;
+      break;
+    default:
+      // Unhandled mouse gesture (like double click/etc).
+      return true;
+  }
+
+  MouseButton pressed_button_mask = MouseButton::kNone;
+  ModifierKey modifier_key_mask = ModifierKey::kNone;
+  if (w_param & MK_LBUTTON) {
+    pressed_button_mask |= MouseButton::kButton1;
+  }
+  if (w_param & MK_MBUTTON) {
+    pressed_button_mask |= MouseButton::kButton2;
+  }
+  if (w_param & MK_RBUTTON) {
+    pressed_button_mask |= MouseButton::kButton3;
+  }
+  if (w_param & MK_XBUTTON1) {
+    pressed_button_mask |= MouseButton::kButton4;
+  }
+  if (w_param & MK_XBUTTON2) {
+    pressed_button_mask |= MouseButton::kButton5;
+  }
+  if (w_param & MK_CONTROL) {
+    modifier_key_mask |= ModifierKey::kCtrl;
+  }
+  if (w_param & MK_SHIFT) {
+    modifier_key_mask |= ModifierKey::kShift;
+  }
+
+  MouseEvent mouse_event{screen_px,     control_px,          wheel_delta,
+                         action_button, pressed_button_mask, modifier_key_mask};
+  switch (message) {
+    case WM_LBUTTONDOWN:
+    case WM_MBUTTONDOWN:
+    case WM_RBUTTONDOWN:
+    case WM_XBUTTONDOWN:
+      PostInputEvent(
+          [mouse_event](InputListener* listener, ref_ptr<Control> control) {
+            listener->OnMouseDown(std::move(control), mouse_event);
+          });
+      break;
+    case WM_LBUTTONUP:
+    case WM_MBUTTONUP:
+    case WM_RBUTTONUP:
+    case WM_XBUTTONUP:
+      PostInputEvent(
+          [mouse_event](InputListener* listener, ref_ptr<Control> control) {
+            listener->OnMouseUp(std::move(control), mouse_event);
+          });
+      break;
+    case WM_MOUSEWHEEL:
+      PostInputEvent(
+          [mouse_event](InputListener* listener, ref_ptr<Control> control) {
+            listener->OnMouseWheel(std::move(control), mouse_event);
+          });
+      break;
+    case WM_MOUSEMOVE:
+      PostInputEvent(
+          [mouse_event](InputListener* listener, ref_ptr<Control> control) {
+            listener->OnMouseMove(std::move(control), mouse_event);
+          });
+      break;
+  }
+
+  // Returning true will prevent default wndproc.
+  return true;
+}
+
+bool Win32Control::HandleKeyboardMessage(UINT message, WPARAM w_param,
+                                         LPARAM l_param) {
+  int key_code = static_cast<int>(w_param);
+
+  ModifierKey modifier_key_mask = ModifierKey::kNone;
+  if ((::GetAsyncKeyState(VK_CONTROL) & 0x8000) == 0x8000) {
+    modifier_key_mask |= ModifierKey::kCtrl;
+  }
+  if ((::GetAsyncKeyState(VK_SHIFT) & 0x8000) == 0x8000) {
+    modifier_key_mask |= ModifierKey::kShift;
+  }
+  if ((::GetAsyncKeyState(VK_MENU) & 0x8000) == 0x8000) {
+    modifier_key_mask |= ModifierKey::kAlt;
+  }
+
+  KeyboardEvent keyboard_event{key_code, modifier_key_mask};
+  switch (message) {
+    case WM_KEYDOWN:
+      DCHECK_LE(key_code, 255);
```

Comment:
yeah - I'd like to know if they are ever hit (as we may want to handle them). will gracefully handle and dcheck above.